openssl-dynamic/src/main/c/sslutils.c

```diff
@@ -188,76 +180,22 @@ void SSL_set_app_data3(SSL *ssl, void *arg)
     SSL_set_ex_data(ssl, SSL_app_data3_idx, arg);
 }
 
-/* Simple echo password prompting */
-int SSL_password_prompt(tcn_pass_cb_t *data)
-{
-    int rv = 0;
-    data->password[0] = '\0';
-    if (data->cb.obj) {
-        JNIEnv *e;
-        jobject  o;
-        jstring  prompt;
-        tcn_get_java_env(&e);
-        prompt = AJP_TO_JSTRING(data->prompt);
-        if ((o = (*e)->CallObjectMethod(e, data->cb.obj,
-                            data->cb.mid[0], prompt))) {
-            TCN_ALLOC_CSTRING(o);
-            if (J2S(o)) {
-                strncpy(data->password, J2S(o), SSL_MAX_PASSWORD_LEN);
-                data->password[SSL_MAX_PASSWORD_LEN-1] = '\0';
-                rv = (int)strlen(data->password);
-            }
-            TCN_FREE_CSTRING(o);
-        }
-    }
-    else {
-#ifdef WIN32
-        rv = WIN32_SSL_password_prompt(data);
-#elif !defined(OPENSSL_IS_BORINGSSL)
-        EVP_read_pw_string(data->password, SSL_MAX_PASSWORD_LEN,
-                           data->prompt, 0);
-#endif
-        rv = (int)strlen(data->password);
-    }
-    if (rv > 0) {
-        /* Remove LF char if present */
-        char *r = strchr(data->password, '\n');
-        if (r) {
-            *r = '\0';
-            rv--;
-        }
-#ifdef WIN32
-        if ((r = strchr(data->password, '\r'))) {
-            *r = '\0';
-            rv--;
-        }
-#endif
-    }
-    return rv;
-}
-
 int SSL_password_callback(char *buf, int bufsiz, int verify,
                           void *cb)
 {
-    tcn_pass_cb_t *cb_data = (tcn_pass_cb_t *)cb;
+    char *password = (char *) cb;
 
-    if (buf == NULL)
+    if (buf == NULL || password == NULL)
         return 0;
     *buf = '\0';
-    if (cb_data == NULL)
-        cb_data = &tcn_password_callback;
-    if (!cb_data->prompt)
-        cb_data->prompt = SSL_DEFAULT_PASS_PROMPT;
-    if (cb_data->password[0]) {
+
+    if (password[0]) {
         /* Return already obtained password */
-        strncpy(buf, cb_data->password, bufsiz);
+        strncpy(buf, password, bufsiz);
         buf[bufsiz - 1] = '\0';
         return (int)strlen(buf);
```

Comment:
also if we expect this to be called frequently should we cache the length of the password in the `tcn_pass_cb_t` struct to avoid calling `strlen` every time.

Also on a related note is it possible this call back will be invoked with an `cb` that points to a `char*` which has already been freed? If it is less likely the `tcn_pass_cb_t` will be freed maybe we should continue passing a pointer to this structure.